cmd-ref: document --rev option on update (#1007)
iterative/dvc.org: public/static/docs/command-reference/update.md

update them.

`dvc update` will not have an effect on import stages that are fixed to a commit
hash (`rev` field in the DVC-file). To update the imported artifacts to a

> certain revision, `--rev` with specified revision can be used.

Reviewer: ```suggestion
`dvc update` without flags will not have an effect on import stages that are
fixed to a commit hash (`rev` field in the DVC-file). Use the `--rev` option
to update an imported artifact to a different commit.
```

```dvc
dvc update --rev master
```

## Options

- `--rev` - specific
  [Git revision](https://git-scm.com/book/en/v2/Git-Internals-Git-References)
  (such as a branch name, a tag, or a commit hash) of the repository to update
  the file or directory from (also starts tracking the given revision).

  > Note that this adds or updates a `rev` field in the DVC-file that fixes it
  > to this revision (and updates `rev_lock` in the DVC-file). This can have an

>   > impact on the behavior of `dvc update` later.

Reviewer: Lets just copy this from import:

```suggestion
- `--rev` - commit hash, branch or tag name, etc. (any
  [Git revision](https://git-scm.com/docs/revisions)) of the repository to
  update the file or directory from. The latest commit in `master` (tip of the
  default branch) is used by default when this option is not specified.

  > Note that this adds a `rev` field in the import stage that fixes it to this
  > revision. This can impact the behavior of `dvc update`. (See
  > **re-importing** example below.)
```

- `-h`, `--help` - prints the usage/help message, and exit.

- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
  problems arise, otherwise 1.

- `-v`, `--verbose` - displays detailed tracing information.

> ## Example: Updating imported artifacts

Reviewer: Just
```suggestion
## Example
```
should be enough here. "Updating imported artifacts" is what this command is about anyway.

> Note that `dvc update` updates the `rev_lock` field of the corresponding
> [DVC-file](/doc/user-guide/dvc-file-format) (when there are changes to bring
> in).

## Example: Updating imported artifacts to a specified revision

> Let's import a data artifact from an older commit from our

Reviewer 2: feels that we now have two similar examples - let's keep one of them and make a link to another?

Author: I am not sure how to do that, as the flow of two examples are different (one starts with --rev, other does not mention it at all).

Reviewer 2: agreed.

Reviewer: It's baiscaly the same as **Example: Fixed revisions and updating to different revision** in `dvc import` though. Maybe we should just copy/paste it or use similar terminology i.e. "fixed revision".

[get started example repo](https://github.com/iterative/example-get-started) at
first:

```dvc

> $ dvc import --rev 72e0f1 git@github.com:iterative/example-get-started model.pkl

Reviewer 2: don't rely on commit hash - they change in that repo

Reviewer 2: split snippet into multiple lines

Reviewer 2: instead of a hash - may be tag is more or less enough - tags might change, but in most cases they are not ... not sure about this

> Importing 'model.pkl (git@github.com:iterative/example-get-started)'

Reviewer 2: no need to include obvious outputs

Reviewer: And we should split the long line:
```suggestion
$ dvc import --rev baseline-experiment \
             git@github.com:iterative/example-get-started \
             model.pkl
```

-> 'model.pkl'
```

After this, the import stage (DVC-file) `model.pkl.dvc` is created.

Let's try to run `dvc update` on the given stage file, and see what happens.

```dvc
$ dvc update model.pkl.dvc
```

There was no output at all, meaning, the `model.pkl` file was not updated. This
is because, we tied the import stage with a `rev` that never changes (i.e.
commit hash never changes). Therefore, it was not updated.

Let's try to update it to the latest version by specifying revision to `master`
branch.

```dvc

> $ dvc update --rev master model.pkl.dvc

Reviewer 2: let's try to not change semantics - from fixed commit to not following the branch

Author: Do you have any good examples in mind? I'm using this as a kind of a hack for being short in examples. I tried using the recent `aita_dataset` but, the dataset on `master` was too large (~117 MB) just for the demonstration (but, it's quite good in that I can change the --rev from `master` to `lightweight`, quite good example really).

Reviewer 2: we can use tags at least? there are two tags in the example-get-started - baseline-experiment and bigrams-experiment

Reviewer 2: tags are more or less static, and you have to use re-import or use `dvc update --rev` to move the latest version

what do you think?

Author: @shcheklein, i moved to tags. Showing example of moving between experiments is good, but, I feel importing from a not-changing revision and trying to move to a recent version of artifacts is also a good example worth showing off.

Reviewer 2: @skshetry agreed, but we need a bit different approach then (we can't rely on a specific hash since it might change). Git clone repo use git to take the latest hash? (basically show how to take it?) .... or even use `dvc list` - may be a good idea for the list to show commit hashes?

Reviewer: Seems like a similar question as https://github.com/iterative/dvc.org/pull/1007#discussion_r386716408:

> It's baiscaly the same as Example: Fixed revisions and updating to different revision in dvc import. Maybe we should just copy/paste it...

> WARNING: DVC-file 'model.pkl.dvc' changed.

Reviewer 2: same here - let's not include outputs

Author: How I wrote the whole examples here, I feel it's important to show the output (that no output is there if there's no any updates). But, we won't need this if we change the example.

Reviewer 2: 👍 up to you on this

> Note that `dvc update` updates the `rev_lock` field of the corresponding
> [DVC-file](/doc/user-guide/dvc-file-format) (when there are changes to bring
> in).

> ## Example: Updating imported artifacts to a specified revision

Reviewer: ```suggestion
## Example: Updating fixed-revision imports to a different version
```
And explain what fixed revision is in the first p. Again, maybe just copy the example from `dvc import`?

Let's import a data artifact from an older commit from our
[get started example repo](https://github.com/iterative/example-get-started) at

> first:

Reviewer: In case we keep this p:
```suggestion
Let's import a model from an old version of our
[get started example repo](https://github.com/iterative/example-get-started)
first:
```

```dvc
$ dvc import --rev baseline-experiment git@github.com:iterative/example-get-started model.pkl
Importing 'model.pkl (git@github.com:iterative/example-get-started)'
-> 'model.pkl'
```

After this, the import stage (DVC-file) `model.pkl.dvc` is created.

> Let's try to run `dvc update` on the given stage file, and see what happens.

Reviewer: ```suggestion
After this, the import stage (DVC-file) `model.pkl.dvc` is created. Let's
try to run `dvc update` on the given stage file, and see what happens.
```

```dvc
$ dvc update model.pkl.dvc
```

There was no output at all, meaning, the `model.pkl` file was not updated. This
is because, we tied the import stage with a `rev` that never changes (i.e. tag

> is tied to a specific commit). Therefore, it was not updated.

Reviewer: ```suggestion
is because we tied the import stage to a `rev` that hasn't changed (i.e.
`baseline-experiment` tag points to a specific Git commit). Therefore, it was not updated.
```
But again, this may be better near the top to explain what a fixed revision is... or just copy the example from `dvc import` instead.

> Let's try to update the model to a different experiment `bigrams-experiment`:

Reviewer: ```suggestion
Let's try to update the model to a different version:
```

```dvc
$ dvc update --rev bigrams-experiment model.pkl.dvc
Importing 'model.pkl (git@github.com:iterative/example-get-started)'

> -> 'model.pkl'

Reviewer: Please remove.

Reviewer 2: Why?

Author: @jorgeorpinel, I wanted to show the outputs here. If there's no output, the data is not updated.

Reviewer: > Why?

Our style from what I remember is to not include output unless there's a good reason and I thought there was no reason. Up to you

> If there's no output, the data is not updated.

On this (out of scope) should `update` print a message when it can't update import stages because they're fixed to a revision, instead of just retuning quietly?

Author: Good point. I'll create a issue for that. We could discuss there.

Reviewer: Thanks. Resolving this one then but if you can link here to the issue that would be great.

> And again @skshetry if you have lots of other work I can take this review over since it the PR is already merged, np just lmk so I don't loose track 🙂

```

The import stage is overwritten, and will get updated from the latest changes in

> the given revision (i.e. `bigrams-experiment` tag).

Reviewer: given commit (tag `bigrams-experiment)

> In the above example, the value for `rev` in the new import stage will be

> > `bigrams-experiment`.

Reviewer: Probably not needed 🙂
